Add Negotiate auth via gssproxy, and password checking via saslauthd

nalind opened the pull request:
These patches add optional use of a gssproxy service to use Negotiate authentication between clients and the apiserver, and teach the apiserver to use saslauthd to check Basic authentication requests.

I figure using a gssproxy would be preferable to linking directly with Kerberos shared libraries, but can rework that if using a cgo binding is preferable.

The changes to the salt configuration could be dropped if need be, though they were time savers for me.

The bits that talk to saslauthd and the change to using the union authenticator can be broken out into a different PR if it makes the review easier.

Thanks!

nalind committed:
Attempt Negotiate auth if Config.GssProxy is set

If Config.GssProxy is set, indicating the location of a gss-proxy socket
to use, attempt to use it to perform Negotiate authentication against
the server, in the style of RFC4559.

The gssproxy service should be up and running, and configured so that
unprivileged clients which connect to it over the specified socket will
be allowed to have it use their Kerberos credentials to perform the
client portion of GSSAPI authentication, like so:

  [service/kube-client]
    mechs = krb5
    socket = /run/gssproxy-kube.sock
    euid = 0
    allow_any_uid = yes
    cred_usage = initiate
    cred_store = ccache:KEYRING:persistent:%U

nalind committed:
Use saslauthd and gssproxy in the apiserver

Add the necessary interfaces and support for accepting Negotiate
authentication from clients by using a gss-proxy service, and for
checking passwords offered via Basic authentication using a saslauthd
service.

When the apiserver is started with -gssproxy=/socket/path, it will
contact a running gss-proxy using the specified socket, and ask it to
perform the server portion of Negotiate authentication.  That gss-proxy
service's configuration should instruct it to use the server's Kerberos
keytab to accept authentication requests from clients, like so:

  [service/kube-apiserver]
    mechs = krb5
    socket = /run/gssproxy-kube-apiserver.sock
    cred_store = keytab:/etc/kube.keytab
    cred_usage = accept
    euid = 0

When the apiserver is started with the -saslauthd flag, which takes
arguments "[realm][,[socket][,[servicename]]]", the server will attempt
to hand Basic credentials off to a running saslauthd service for
checking.  The default socket location, if none or an empty value is
specified, is "/var/run/saslauthd/mux".  The default service name, if
none or an empty value is specified, is based on the name of the binary
("kube-apiserver", typically).

On most Fedora or Red Hat systems, saslauthd is configured to use PAM by
default, so either the service name should be set to a valid PAM
configuration ("password-auth"), or a configuration should be created
for the name which the server will use.

The server will accept any of bearer tokens, negotiate, or basic auth
now, and requires that at least one of them be configured.

nalind committed:
Enable saslauthd and gssproxy in salt

Update the salt provisioning so that we install and enable gssproxy on
both masters and minions, and install and enable saslauthd on the
master.  Currently we mark this as specific to Red Hat systems (and
there are some kinks to work out yet) to avoid breaking things due to
possible differences in packaging of gssproxy and cyrus-sasl which we
haven't accounted for.

nalind commented:
To be clear, you're referring to the part that touches the Godeps configuration, or something else?

smarterclayton commented:
@liggitt review please

smarterclayton commented:
@nalind yeah, anything that involves the Godeps/* dir should be its own commit, and try to give the commit name the packages you're updating and commits.  It's in version control, but it's an easy way for someone reviewing to double check and scan.

nalind commented:
Ah, okay.  Will do.

nalind commented:
The client option might be a little much.  I haven't sorted out advertising what the server supports in that initial 401 response, but if that works, then a) browsers would know what to do, and b) we could set a reasonable default on the client side.

nalind committed:
bump(github.com/davecgh/go-xdr/xdr2):95e2c801038137f3f52d9f0cf351121faf005f27

nalind committed:
bump(github.com/nalind/gss/pkg/gss/proxy):4a6dff9e91f39dea723a3ee7b125ab04c2888e7e

liggitt commented on cmd/kube-apiserver/apiserver.go:
```diff
@@ -165,9 +170,29 @@ func main() {
+		}
+		authenticators = append(authenticators, authenticator)
+	}
+	if len(authenticators) == 0 {
```
authentication isn't required today, so don't fail on an empty authenticators list

nalind replied:
Fixing.

liggitt commented on pkg/auth/authenticator/basic/basic.go:
```diff
@@ -0,0 +1,48 @@
+package basic
```
There is already a basicauth authenticator in plugin/pkg/auth/authenticator/request/basicauth

nalind replied:
Doh.  Dropping.

liggitt commented on pkg/auth/authenticator/interfaces.go:
```diff
@@ -43,6 +43,11 @@ type Password interface {
 }
 
+// Negotiate checks a context initiation token using a GssProxy.
+type Negotiate interface {
```
Can you use the existing Token interface?

nalind replied:
Currently it expects the client's authorization header to be marked as "bearer".  Any preference on how it should be made to accept "negotiate"?

nalind replied:
Never mind, I was getting that confused with the bearer implementation.

liggitt commented on pkg/auth/authenticator/negotiate/negotiate.go:
```diff
@@ -0,0 +1,37 @@
+package negotiate
```
Move to a plugin under plugin/pkg/auth/authenticator/request/negotiate

nalind replied:
Moving.

liggitt commented on plugin/pkg/auth/authenticator/gssapi/gssproxy/gssproxy.go:
```diff
@@ -0,0 +1,73 @@
+	}
+	if ascr.Status.MajorStatus == proxy.S_CONTINUE_NEEDED {
+		// We can't handle multiple-round-trip requests right.
+		return nil, false, errors.New("Negotiate authentication canceled (incomplete): " + ascr.Status.MajorStatusString + "/" + ascr.Status.MinorStatusString)
```
I thought all negotiate interchanges were round-trip, initiated by a WWW-Authenticate: Negotiate challenge from the host. Is that not the case?

nalind replied:
It's what I'd prefer, but the last paragraph of section 4.2 in RFC4559 indicates that it's not required, and 2617 says that it's usual, but not necessary.  I can't find where we set an authorization header in the 401 response.

nalind replied:
I've got additional changes to add WWW-Authenticate headers for Basic/Bearer/Negotiate when the server's configured to use them.  Do you have a preference between adding them to this PR, or saving them for another one to be opened after this one is resolved?

liggitt replied:
I'm still not convinced we want kerberos built into the kubernetes server. That feels like something you'd be more likely to add as an auth proxy (not to mention the auth interfaces aren't set up for round-trip negotiates).

nalind replied:
The fixes for setting WWW-Authenticate headers will add a way to send a Negotiate response back to the client, though the client's already verified the server's identity using its certificate, so requesting mutual auth from GSSAPI is arguably redundant.

We're handing the Negotiate data off to a gss-proxy daemon to do the heavy lifting, largely to avoid having to link directly with any Kerberos or GSSAPI shared libraries.  Can you elaborate on where an auth proxy would be wired in?

liggitt replied:
As a front proxy (like Apache running mod_auth_kerb), so that any requests that made it through to the API would have already been kerberos-authenticated. Even non-mutual auth cases require round-tripping if the client waits for a WWW-Authenticate: Negotiate challenge before sending an Authorization header

nalind replied:
Right, but kubectl isn't the type of client to do that, since its authentication configuration's pre-specified.  That's more of a browser (and when I tried it, curl) case.

Being able to accept auth via a proxy doesn't seem to require that the server be unable to handle some things by itself.  We'd need an additional plugin to accept user identity information provided by a proxy, though.

nalind replied:
The patches for the client and the server should be cleanly separated, so if it's preferable that we only consider the client bits, the bits that touch the server could be dropped or deferred.

liggitt commented on plugin/pkg/auth/authenticator/gssapi/gssproxy/gssproxy.go:
```diff
@@ -0,0 +1,73 @@
+		return nil, false, errors.New("Negotiate authentication failed: " + ascr.Status.MajorStatusString + "/" + ascr.Status.MinorStatusString)
+	}
+	// We also don't have a method for returning ascr.OutputToken to the client yet.
+	return &user.DefaultInfo{Name: secCtx.SrcName.DisplayName}, true, nil
```
Define an interface that transforms the secCtx (or SrcName, not sure which) into a user.DefaultInfo (see x509.UserConversion for an example). The default implementation can do more or less what you do here, though I think we'd prefer the id over the displayname if that's available

nalind replied:
Okay, I think I've got what you're looking for in there now.

liggitt commented on plugin/pkg/auth/authenticator/password/saslauthd/saslauthd.go:
```diff
@@ -0,0 +1,144 @@
+		username = username[0:index]
+	}
+	if username == "" {
+		return nil, false, errors.New("No user name specified for basic authentication")
```
is a blank username/password an error condition, or should it just return `nil, false, nil`?

nalind replied:
If I'm reading RFC2617 right, it's technically allowed, so I'm dropping that check.

liggitt commented on plugin/pkg/auth/authenticator/password/saslauthd/saslauthd.go:
```diff
@@ -0,0 +1,144 @@
+		return nil, false, errors.New("No user name specified for basic authentication")
+	}
+	if password == "" {
+		return nil, false, errors.New("No password specified for basic authentication")
```
would an empty password ever be allowed? if so, this code shouldn't decide it's an error, but should let the sasl service fail it

nalind replied:
I guess it could be.  PAM certainly allows it.  Removing the check.

nalind committed:
Return challenges in 401 responses

Allow authentication plugins to provide HTTP headers to be sent to a
client when they're called to verify authorization headers.  The union
authenticator will return all headers from all of the plugins which it
calls if none succeed, but only the headers from a succeeding module if
one does.

Add logic to provide authentication challenge headers when they're
required but the client has not provided one.  Plugins which check
passwords for Basic authentication now need to provide a GetRealm()
function which returns the Basic challenge's "realm" parameter.

While kubectl doesn't currently use this information, other clients
(some browsers, curl) expect to see a challenge header before they'll
attempt authentication using a given mechanism.

nalind commented:
Rebased to fix a merge conflict, tweaked an error message.

liggitt commented:
@nalind I'm still not convinced we want kerberos built into the kubernetes server... that feels like something you would be more likely to add via an auth proxy.

nalind committed:
Run gendocs

nalind commented:
This should be covered by Red Hat's CLA.

erictune commented:
ping @smarterclayton @nalind what is the status of this PR?  abandoned, or going to come back to it?

piosz commented:
Is this PR active? If not please close it according to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/pull-requests.md

piosz commented:
Closing according to the mentioned policy.
